builtin/common/vector.lua

```diff
@@ -1,58 +1,96 @@
+--[[
+Vector helpers
+
+Note: The functions must be able to accept old vectors that had no metatables and
+      looked like this: {x = 42, y = 13, z = 0}
+]]
 
 vector = {}
 
+local metatable = {}
+
+function metatable.__index(v, key)
+	if type(key) == "number" then
+		return v[key]
+	elseif key == "x" then
+		return v[1]
+	elseif key == "y" then
+		return v[2]
+	elseif key == "z" then
+		return v[3]
+	end
+	local func = vector[key]
+	if func then
+		return func
+	else
+		error("Invalid key for vector", 2)
+	end
+end
+
+function metatable.__newindex(v, key, value)
+	if type(key) == "number" then
+		v[key] = value
+	elseif key == "x" then
+		v[1] = value
+	elseif key == "y" then
+		v[2] = value
+	elseif key == "z" then
+		v[3] = value
+	else
+		error("Invalid key for vector", 2)
+	end
+end
+
 function vector.new(a, b, c)
 	if type(a) == "table" then
 		assert(a.x and a.y and a.z, "Invalid vector passed to vector.new()")
-		return {x=a.x, y=a.y, z=a.z}
+		return setmetatable({a.x, a.y, a.z}, metatable)
 	elseif a then
 		assert(b and c, "Invalid arguments for vector.new()")
-		return {x=a, y=b, z=c}
+		return setmetatable({a, b, c}, metatable)
 	end
-	return {x=0, y=0, z=0}
+	return setmetatable({0, 0, 0}, metatable)
 end
 
 function vector.equals(a, b)
 	return a.x == b.x and
 	       a.y == b.y and
 	       a.z == b.z
 end
+metatable.__eq = vector.equals
 
 function vector.length(v)
 	return math.hypot(v.x, math.hypot(v.y, v.z))
 end
+-- Note: we can not use __len because it is already used for primitive table length
 
 function vector.normalize(v)
 	local len = vector.length(v)
 	if len == 0 then
-		return {x=0, y=0, z=0}
+		return vector.new()
 	else
 		return vector.divide(v, len)
 	end
 end
 
 function vector.floor(v)
-	return {
-		x = math.floor(v.x),
-		y = math.floor(v.y),
-		z = math.floor(v.z)
-	}
+	return vector.apply(v, math.floor)
 end
 
 function vector.round(v)
-	return {
-		x = math.floor(v.x + 0.5),
-		y = math.floor(v.y + 0.5),
-		z = math.floor(v.z + 0.5)
-	}
+	return vector.new(
+		math.floor(v.x + 0.5),
+		math.floor(v.y + 0.5),
+		math.floor(v.z + 0.5)
+	)
```

Comment:
I worry a little about `vector.new` (... if-else and asserts) being called for every calculation

Reply:
Done.